### PR TITLE
Add Convex triggers

### PR DESCRIPTION
**Pull request by mds1**

### CODE REVIEW

File: test/Convex.test.ts [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+          // Decrease virtual price by an amount more than tolerance, should be triggered
+          await modifyLastVirtualPrice(tolerance - 1n, 1000n);
+          await assertTriggerStatus(true);
+        });
+
+        // Due to a bug in Smock (https://github.com/defi-wonderland/smock/issues/101) we leave these tests
+        // skipped because they cause other tests to fail. Replace the `.skip` with a `.only` to run these
+        // tests and verify that they pass
+        it.skip(`toggles trigger when ${poolType} pool's get_virtual_price() reverts`, async () => {
```
Comment (apbendi) [inline test/Convex.test.ts:275]: Random thought: if you manually called setupFixture here, effectively deploying a new set of contracts for use only with this test, would it work around this issue allowing the test to be included in the suite without skipping?

File: contracts/Convex.sol [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    baseToken1 = basePoolCoins(1);
+    baseToken2 = basePoolCoins(2);
+
+    // Get virtual prices
+    lastVpMetaPool = ICrvMeta(curveMetaPool).get_virtual_price();
+    lastVpBasePool = ICrvBase(curveBasePool).get_virtual_price();
+  }
+
+  function checkTriggerCondition() internal override returns (bool) {
+    // Typically in this method we check all conditions, save them to storage, amd return the result.
```
Comment (apbendi) [inline contracts/Convex.sol]: nit: typo "amd"

File: contracts/Convex.sol [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    // This is convenient because it ensures we have the data that caused the trigger saved into
+    // the state, but this is just convenient and not a requirement. We do not follow that pattern
+    // here because  certain trigger conditions can cause this method to revert if we tried that
```
Comment (apbendi) [inline contracts/Convex.sol]: nit: extra space between "because" and "certain"

File: contracts/Convex.sol [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+    // Verify supply of Convex receipt tokens is equal to the amount of curve receipt tokens Convex
+    // can claim. Convex receipt tokens are minted 1:1 with deposited funds, so this protects
+    // against e.g. "infinite mint" type bugs, where an attacker is able to mint themselves more
+    // Convex receipt tokens than what they should receive.
+    if (IERC20(convexToken).totalSupply() != IERC20(gauge).balanceOf(staker)) return true;
+
+    // Token balance checks
+    if (checkCurveBaseBalances()) return true;
+    if (checkCurveMetaBalances()) return true;
```
Comment (apbendi) [inline contracts/Convex.sol]: nit: could be one `if` with a `||`. Depending on how smart the compiler is, might be a minor gas optimization.

**Review by apbendi — APPROVED** [review]
A few minor comments, but otherwise looks good to me and tests pass locally.

**Review by lvwrence — APPROVED** [review]
lgtm

File: test/Convex.test.ts (hunk 1 above)
Comment (mds1) [reply test/Convex.test.ts:275]: Ah good idea. It sounds like it should work, but in my quick tests it did not, so leaving this as-is for now in the interest of time 😕

File: contracts/Convex.sol (hunk 2 above)
Comment (mds1) [reply contracts/Convex.sol]: Fixed in fd73021

File: contracts/Convex.sol (hunk 3 above)
Comment (mds1) [reply contracts/Convex.sol]: Fixed in fd73021

File: contracts/Convex.sol (hunk 4 above)
Comment (mds1) [reply contracts/Convex.sol]: Changed in fd73021

**Pull request merged**